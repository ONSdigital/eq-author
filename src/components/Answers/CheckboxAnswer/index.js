import React from "react";
import PropTypes from "prop-types";

import { TransitionGroup, CSSTransition } from "react-transition-group";

import Button from "../../Button";
import { Field, Input } from "../../Forms";

import SeamlessInput from "../../QuestionnaireDesign/SeamlessInput";
import SeamlessTextArea from "../../QuestionnaireDesign/SeamlessTextArea";

import styled from "styled-components";

const duration = 200;

const CheckboxAnswerWrapper = styled.div`
  width: 20em;
  margin: 0 auto;
`;

const CheckboxOptions = styled.div`margin: 1em 0;`;

export const CheckboxOption = styled.div`
  border: 1px solid #e5e5e5;
  padding: 1em 1em 0em 1em;
  border-radius: 4px;

  position: relative;

  transition: height ${duration / 2}ms ease-out,
    opacity ${duration}ms ease-out ${duration}ms, padding ${duration}ms ease-out,
    transform ${duration}ms ease-out ${duration}ms;
  opacity: 1;
  transform: translateX(0);

  &:not(:first-child) {
    margin-top: .5em;
  }

  &.option-enter {
    opacity: 0;
    height: 0;
    padding: 0;
    transform: translateX(-20px);
  }

  &.option-entered {
    height: auto;
  }
`;

const SeamlessLabel = styled(SeamlessInput)`
  display: inline-block !important;
  width: auto;
`;

export const AddOtherLink = styled.a`
  color: #48a6f6;
  text-decoration: none;
`;

const StyledCheckboxInput = styled(Input)`
  border: 2px solid #979797;
  height: 1.4em;
  width: 1.4em;
`;

export const CloseButton = styled.button`
  float: right;
  cursor: pointer;
  color: #ccc;
  padding: .2em;
  border: 0;
  background: none;
  font-size: 1em;

  position: absolute;
  top: .5em;
  right: 1em;

  transition: color .2s ease-in-out;

  &:hover {
    color: #333;
    transition: color .2s ease-in-out;
  }
`;

const OtherOption = styled.small`margin: 0 .5em;`;

const CheckboxAnswer = ({
  options,
  onChangeLabel,
  onChangeDescription,
  onAddOption,
  onDeleteOption,
  onAddOther
}) => {
  return (
    <CheckboxAnswerWrapper>
      <TransitionGroup component={CheckboxOptions}>
        {options &&
          options.map(option =>
            <CSSTransition
              key={option.id}
              timeout={duration}
              classNames="option"
            >
              <CheckboxOption key={option.id}>
                <Field id={"option." + option.id + ".label"}>
                  <StyledCheckboxInput type="checkbox" disabled />
                  <SeamlessLabel
                    placeholder="Label"
                    size="medium"
                    onChange={onChangeLabel}
                    value={option.label}
                  />
                </Field>
                <Field id={"option." + option.id + ".description"}>
                  <SeamlessTextArea
                    cols="30"
                    rows="5"
                    placeholder="Optional description"
                    onChange={onChangeDescription}
                    value={option.description}
                  />
                </Field>
                <CloseButton
                  name={"option." + option.id + ".delete"}
                  onClick={onDeleteOption}
                >
                  &times;
                </CloseButton>
              </CheckboxOption>
            </CSSTransition>
          )}
      </TransitionGroup>
      <div>
        <Button secondary onClick={onAddOption}>
          Add another option
        </Button>
        <OtherOption>
          or{" "}
          <AddOtherLink href="#" onClick={onAddOther}>
            add other
          </AddOtherLink>
        </OtherOption>
      </div>
    </CheckboxAnswerWrapper>
  );
};

CheckboxAnswer.defaultProps = {
  options: [
    {
      id: 1,
      label: "",
      description: ""
    }
  ]
};

CheckboxAnswer.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      label: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired
    })
  ),
  onChangeLabel: PropTypes.func.isRequired,
  onChangeDescription: PropTypes.func.isRequired,
  onAddOption: PropTypes.func.isRequired,
  onDeleteOption: PropTypes.func.isRequired,
  onAddOther: PropTypes.func.isRequired
};

export default CheckboxAnswer;
