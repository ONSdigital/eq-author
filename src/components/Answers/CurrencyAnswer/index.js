import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { colors, radius } from "constants/theme";
import BasicAnswer from "components/Answers/BasicAnswer";
import DummyTextInput from "components/Answers/Dummy/TextInput";

const StyledSpan = styled.span`
  display: inline-block;
  background-color: #f5f5f5;
  border-right: 1px solid ${colors.borders};
  border-radius: ${radius} 0 0 ${radius};
  padding: 0.6em 0;
  width: 2.5em;
  font-weight: 700;
  font-size: 1em;
  text-align: center;
  position: absolute;
  left: 0;
  top: 0;
  z-index: 1;
`;

const FieldWrapper = styled.div`
  display: block;
  width: 100%;
  margin-bottom: 1em;
  position: relative;
`;

const CurrencyComponent = props =>
  <StyledSpan>
    {props.currencyUnit}
  </StyledSpan>;

CurrencyComponent.propTypes = {
  currencyUnit: PropTypes.string
};

CurrencyComponent.defaultProps = {
  currencyUnit: "Â£"
};

const CurrencyAnswer = props =>
  <BasicAnswer {...props}>
    <FieldWrapper>
      <CurrencyComponent />
      <DummyTextInput />
    </FieldWrapper>
  </BasicAnswer>;

export default CurrencyAnswer;
