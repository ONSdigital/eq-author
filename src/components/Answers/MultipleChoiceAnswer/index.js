import React, { Component } from "react";

import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import { TransitionGroup } from "react-transition-group";
import styled from "styled-components";
import focusOnEntity from "utils/focusOnEntity";

import Option from "./Option";
import OptionTransition from "./OptionTransition";
import BasicAnswer from "components/Answers/BasicAnswer";

import { colors } from "constants/theme";
import { CHECKBOX } from "constants/answer-types";
import SplitButton from "components/SplitButton";
import Dropdown from "components/SplitButton/Dropdown";
import MenuItem from "components/SplitButton/MenuItem";

import { last } from "lodash";
import gql from "graphql-tag";

const AnswerWrapper = styled.div`
  margin: 2em 0 0;
  width: 75%;
`;

const AnswerHelper = styled.div`
  margin-bottom: 0.5em;
  font-size: 0.9em;
  font-weight: 600;
  color: ${colors.text};
`;

const Options = styled.div`
  margin: 0 0 1em;
`;

export const AddOtherLink = styled.button`
  color: #48a6f6;
  text-decoration: none;
  border: 0;
  background: none;
`;

const OtherAnswerWrapper = styled.div`
  padding-top: 0.25em;
  margin-bottom: 1em;
`;

class MultipleChoiceAnswer extends Component {
  static propTypes = {
    answer: CustomPropTypes.answer.isRequired,
    onUpdate: PropTypes.func.isRequired,
    onAddOption: PropTypes.func.isRequired,
    onUpdateOption: PropTypes.func.isRequired,
    onDeleteOption: PropTypes.func.isRequired,
    onAddOther: PropTypes.func.isRequired,
    onDeleteOther: PropTypes.func.isRequired,
    minOptions: PropTypes.number.isRequired
  };

  static defaultProps = {
    minOptions: 1
  };

  state = {
    open: false
  };

  handleOptionDelete = optionId => {
    this.props.onDeleteOption(optionId, this.props.answer.id);
  };

  handleAddOption = e => {
    e.preventDefault();
    e.stopPropagation();
    return this.props.onAddOption(this.props.answer.id).then(focusOnEntity);
  };

  handleAddOther = e => {
    e.preventDefault();
    return this.props
      .onAddOther(this.props.answer)
      .then(res => {
        this.handleToggleOpen(false);
        return res;
      })
      .then(res => res.option)
      .then(focusOnEntity);
  };

  handleDeleteOther = e => {
    return this.props
      .onDeleteOther(this.props.answer)
      .then(() => last(this.props.answer.options))
      .then(focusOnEntity);
  };

  handleToggleOpen = open => {
    this.setState({
      open
    });
  };

  render() {
    const {
      answer,
      onUpdateOption,
      onUpdate,
      minOptions,
      ...otherProps
    } = this.props;

    const numberOfOptions = answer.options.length + (answer.other ? 1 : 0);
    const showDeleteOption = numberOfOptions > minOptions;

    return (
      <BasicAnswer
        answer={answer}
        onUpdate={onUpdate}
        autoFocus={false}
        labelText="Label (optional)"
      >
        <AnswerWrapper>
          {answer.type === CHECKBOX && (
            <AnswerHelper>Select all that apply</AnswerHelper>
          )}
          <TransitionGroup component={Options}>
            {answer.options.map((option, optionIndex, options) => (
              <OptionTransition key={option.id}>
                <Option
                  {...otherProps}
                  option={option}
                  onDelete={this.handleOptionDelete}
                  onUpdate={onUpdateOption}
                  onEnterKey={this.handleAddOption}
                  hasDeleteButton={showDeleteOption}
                />
              </OptionTransition>
            ))}
            {answer.other && (
              <OptionTransition key={answer.other.option.id}>
                <Option
                  {...otherProps}
                  option={answer.other.option}
                  onDelete={this.handleDeleteOther}
                  onUpdate={onUpdateOption}
                  onEnterKey={this.handleAddOption}
                  hasDeleteButton={showDeleteOption}
                  labelPlaceholder="eg. Other"
                >
                  <OtherAnswerWrapper data-test="other-answer">
                    <BasicAnswer
                      answer={answer.other.answer}
                      onUpdate={onUpdate}
                      showDescription={false}
                      labelText="Other label"
                      labelPlaceholder="eg. Please specify"
                      bold={false}
                    />
                  </OtherAnswerWrapper>
                </Option>
              </OptionTransition>
            )}
          </TransitionGroup>
          <div>
            <SplitButton
              onPrimaryAction={this.handleAddOption}
              primaryText="Add another option"
              onToggleOpen={this.handleToggleOpen}
              open={this.state.open}
              dataTest="btn-add-option"
            >
              <Dropdown>
                <MenuItem
                  onClick={this.handleAddOther}
                  disabled={answer.other !== null}
                  data-test="btn-add-option-other"
                >
                  Add &quot;other&quot; option
                </MenuItem>
              </Dropdown>
            </SplitButton>
          </div>
        </AnswerWrapper>
      </BasicAnswer>
    );
  }
}

MultipleChoiceAnswer.fragments = {
  MultipleChoice: gql`
    fragment MultipleChoice on Answer {
      ...Answer
      ... on MultipleChoiceAnswer {
        options {
          ...Option
        }
        other {
          option {
            ...Option
          }
          answer {
            ...Answer
          }
        }
      }
    }
    ${Option.fragments.Option}
    ${BasicAnswer.fragments.Answer}
  `
};

export default MultipleChoiceAnswer;
