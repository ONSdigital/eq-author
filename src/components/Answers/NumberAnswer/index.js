import React from "react";
import styled from "styled-components";
import BasicAnswer from "components/Answers/BasicAnswer";
import DummyTextInput from "components/Answers/Dummy/TextInput";

const Width = styled.div`
  position: relative;
  width: 50%;
`;

const NumberAnswer = props => (
  <BasicAnswer {...props}>
    <Width>
      <DummyTextInput placeholder="eg. 123" />
    </Width>
  </BasicAnswer>
);

export default NumberAnswer;
