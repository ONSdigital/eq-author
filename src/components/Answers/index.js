import React from 'react'
import Answer from 'components/Answer'

const Answers = ({answers, ...other}) => (
  <fieldset className="answer-list">
    <legend>Answers</legend>
    { answers.map((answer, index) => (
      <Answer answer={answer} key={index} answerIndex={index} {...other} />
    ))}
  </fieldset>
)
export default Answers
