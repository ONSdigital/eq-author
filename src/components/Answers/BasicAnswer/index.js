import React from "react";
import { Field } from "components/Forms";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import SeamlessTextArea from "components/SeamlessTextArea/SeamlessTextArea";
import withEntityEditor from "components/withEntityEditor";
import answerFragment from "graphql/fragments/answer.graphql";

export const StatelessBasicAnswer = ({
  answer,
  onChange,
  onUpdate,
  children,
  id
}) => (
  <div id={id}>
    <Field id="label">
      <SeamlessTextArea
        placeholder="Label"
        size="medium"
        onChange={onChange}
        onBlur={onUpdate}
        value={answer.label}
        data-autofocus
      />
    </Field>
    <Field id="description">
      <SeamlessTextArea
        cols="30"
        rows="5"
        placeholder="Enter a description (optional)â€¦"
        onChange={onChange}
        onBlur={onUpdate}
        value={answer.description}
      />
    </Field>
    {children}
  </div>
);

StatelessBasicAnswer.propTypes = {
  answer: CustomPropTypes.answer.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
  children: PropTypes.element.isRequired,
  id: PropTypes.string
};

export default withEntityEditor("answer", answerFragment)(StatelessBasicAnswer);
