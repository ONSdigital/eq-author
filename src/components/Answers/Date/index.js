import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";

import DummyDate from "components/Answers/Dummy/Date";
import styled from "styled-components";

import { Field } from "components/Forms";
import WrappingInput from "components/WrappingInput";
import withEntityEditor from "components/withEntityEditor";
import answerFragment from "graphql/fragments/answer.graphql";

const Wrapper = styled.div.attrs({
  role: "presentation"
})`
  cursor: default;
  user-select: none;
  margin-bottom: 0.5em;

  &:last-of-type {
    margin-bottom: 0;
  }
`;

const Date = ({ answer, onChange, onUpdate }) => (
  <Wrapper id={`data-${answer.id}`}>
    <Field>
      <WrappingInput
        name="label"
        placeholder="Label"
        size="medium"
        onChange={onChange}
        onBlur={onUpdate}
        value={answer.label}
        data-autofocus
        data-test="date-answer-label"
      />
    </Field>
    <DummyDate />
  </Wrapper>
);

Date.propTypes = {
  answer: CustomPropTypes.answer.isRequired,
  onChange: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired
};

export default withEntityEditor("answer", answerFragment)(Date);
