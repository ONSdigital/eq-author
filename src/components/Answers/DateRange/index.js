import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import Date from "../Date";

const DateRange = ({ answer, ...otherProps }) =>
  answer.childAnswers.map(answer => (
    <Date key={answer.id} answer={answer} {...otherProps} />
  ));

DateRange.propTypes = {
  answer: CustomPropTypes.answer.isRequired,
  onUpdate: PropTypes.func.isRequired
};

export default DateRange;
