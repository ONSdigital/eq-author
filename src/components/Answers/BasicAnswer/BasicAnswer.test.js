import React from "react";
import { shallow, mount } from "enzyme";
import { StatelessBasicAnswer } from "components/Answers/BasicAnswer";
import WrappingInput from "components/WrappingInput";

describe("BasicAnswer", () => {
  let answer;
  let onChange;
  let onUpdate;
  let children;
  let props;

  const createWrapper = (props, render = shallow) => {
    return render(<StatelessBasicAnswer {...props} />);
  };

  beforeEach(() => {
    answer = {
      title: "Answer title",
      description: "Answer description",
      label: "",
      type: "TextField"
    };
    onChange = jest.fn();
    onUpdate = jest.fn();

    props = {
      id: "1",
      answer,
      onChange,
      onUpdate,
      children: <div>This is the child component</div>
    };
  });

  it("should render", () => {
    expect(createWrapper(props, children)).toMatchSnapshot();
  });

  it("can turn off auto-focus", () => {
    let wrapper = createWrapper({ ...props, autoFocus: false }, mount);
    const input = wrapper
      .find(`[data-test="txt-answer-label"]`)
      .first()
      .getDOMNode();

    expect(input.hasAttribute("data-autofocus")).toBe(false);
  });

  describe("event handling behaviour", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = createWrapper(props, children, mount);
    });

    it("should invoke update callback onBlur", () => {
      wrapper.find(WrappingInput).forEach(input => input.simulate("blur"));

      expect(onUpdate).toHaveBeenCalledTimes(2);
    });

    it("should invoke change callback onChange", () => {
      wrapper.find(WrappingInput).forEach(input => input.simulate("change"));

      expect(onChange).toHaveBeenCalledTimes(2);
    });
  });
});
