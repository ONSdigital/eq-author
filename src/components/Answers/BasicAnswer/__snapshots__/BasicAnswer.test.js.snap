// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicAnswer should render 1`] = `
<div>
  <Field
    last={false}
  >
    <Label
      htmlFor="answer-label-undefined"
    >
      Label
    </Label>
    <withChangeHandler(WrappingInput)
      bold={true}
      data-autofocus={true}
      data-test="txt-answer-label"
      id="answer-label-undefined"
      name="label"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </Field>
  <Field
    last={false}
  >
    <Label
      htmlFor="answer-description-undefined"
    >
      Description (optional)
    </Label>
    <withChangeHandler(WrappingInput)
      cols="30"
      data-test="txt-answer-description"
      id="answer-description-undefined"
      name="description"
      onBlur={[Function]}
      onChange={[Function]}
      rows="5"
      value="Answer description"
    />
  </Field>
  <div>
    This is the child component
  </div>
</div>
`;
